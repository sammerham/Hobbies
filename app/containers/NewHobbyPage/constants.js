export const ADD_HOBBY_ASYNC = 'ADD_HOBBY_ASYNC';
export const ADD_HOBBY = 'ADD_HOBBY';
export const ADD_REQUEST_FAILED = 'ADD_REQUEST_FAILED';
