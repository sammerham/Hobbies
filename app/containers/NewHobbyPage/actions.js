/*
 *
 * NewHobbyPage actions
 *
 */

import { ADD_HOBBY } from './constants';
// dispatched when hobby added
export function addHobby(hobby) {
  return {
    type: ADD_HOBBY,
    hobby,
  };
}
