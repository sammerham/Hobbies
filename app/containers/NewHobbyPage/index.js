import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
// import { v4 as uuidv4 } from 'uuid';
import { useHistory } from 'react-router-dom';
import { ADD_HOBBY } from '../../constants/constants';
import {
  StyledTitle,
  StyledButton,
  StyledForm,
} from '../../styledComponents/styledComponents';

/*
 * NewHobbyPage
 *
 * This is the page with a form to add a new hobby, at the '/new' route
 * State: formData
 * Routes-- -> NewHobbyPage
 */

export default function NewHobbyPage() {
  const dispatch = useDispatch();
  const history = useHistory();
  const [formData, setFormData] = useState({ hobby: '' });

  // handles submit and redirects to homepage on success
  function handleSubmit(evt) {
    evt.preventDefault();
    // dispatch({ type: ADD_HOBBY, hobby: formData.hobby, id: uuidv4() });
    dispatch({ type: ADD_HOBBY, hobby: formData.hobby });
    setFormData({ hobby: '' });
    history.push('/');
  }

  // handle chnage fn for the form
  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData(data => ({ ...data, [name]: value }));
  }

  return (
    <div>
      <StyledTitle bg="dimgray">New Hobby</StyledTitle>
      <StyledForm onSubmit={handleSubmit}>
        <input
          name="hobby"
          id="hobby"
          aria-label="hobby"
          required
          placeholder="Enter a new hobby"
          value={formData.hobby}
          onChange={handleChange}
        />
        <StyledButton bg="dimgray">Add</StyledButton>
      </StyledForm>
    </div>
  );
}
