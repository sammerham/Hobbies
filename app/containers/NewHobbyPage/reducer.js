import { ADD_HOBBY_ASYNC, ADD_REQUEST_FAILED } from './constants';

const DEFAULT_STATE = {
  hobbies: [],
  error: false,
};

export default function addHobbyReducer(state = DEFAULT_STATE, action) {
  switch (action.type) {
    case ADD_HOBBY_ASYNC:
      return { hobbies: action.hobbies };
    case ADD_REQUEST_FAILED:
      return { error: action.error };
    default:
      return state;
  }
}
