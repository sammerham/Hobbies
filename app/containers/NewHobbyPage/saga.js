import { put, takeEvery, call } from 'redux-saga/effects';
import axios from 'axios';
import { addHobbyAsync, addRequestFailed } from './actions';
import { ADD_HOBBY } from './constants';

// worker saga: makes the api call when watcher saga sees the ADD_HOBBY action
function* addHobbiessAsync(action) {
  // post a hobby to api
  /** 
   * @params : none
   * @returns 
  {
    "hobbies":
    [
    "swimming",
    "running",
    "reading",
    "walking",
    "gardening"
  ]
}
*/
  try {
    // calls api to post a hobby to hobbies
    const result = yield call(axios.post, '/api', {
      hobby: action.hobby,
    });
    const { hobbies } = result.data;
    // dispatches ADD_HOBBY_ASYNC action to the store with hobbies
    yield put(addHobbyAsync(hobbies));
  } catch (error) {
    // dispatches ADD_REQUEST_FAILED action to the store with the error
    yield put(addRequestFailed(error));
  }
}

// watcher saga: watches for ADD_HOBBY action dispatched to the store, starts worker saga
export function* watchAddHobbiesAsync() {
  yield takeEvery(ADD_HOBBY, addHobbiessAsync);
}
