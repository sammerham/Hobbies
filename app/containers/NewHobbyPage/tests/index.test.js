import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import { memoryHistory } from 'react-router-dom';
import { Provider } from 'react-redux';
import { IntlProvider } from 'react-intl';
import configureStore from '../../../configureStore';

import NewHobbyPage from '../index';

const store = configureStore({}, memoryHistory);

describe('<NewHobbyPage />', () => {
  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <IntlProvider locale="en">
        <Provider store={store}>
          <NewHobbyPage />
        </Provider>
      </IntlProvider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('should display input field value', () => {
    const { getByPlaceholderText } = render(
      <IntlProvider locale="en">
        <Provider store={store}>
          <NewHobbyPage />
        </Provider>
      </IntlProvider>,
    );
    const input = getByPlaceholderText('Enter a new hobby');
    fireEvent.change(input, { target: { value: 'test hobby' } });
    expect(input.value).toBe('test hobby');
  });
});
