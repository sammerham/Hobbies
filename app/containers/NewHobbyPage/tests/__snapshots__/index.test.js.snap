// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewHobbyPage /> should render and match the snapshot 1`] = `
<div>
  <h1
    class="styledComponents__StyledTitle-by26az-0 cqdktE"
  >
    New Hobby
  </h1>
  <form
    class="styledComponents__StyledForm-by26az-4 lcZgNs"
  >
    <input
      aria-label="hobby"
      id="hobby"
      name="hobby"
      placeholder="Enter a new hobby"
      required=""
      value=""
    />
    <button
      class="styledComponents__StyledButton-by26az-1 iyphPq"
    >
      Add
    </button>
  </form>
</div>
`;
